WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+instance with some action, such as in worklets or with custom elements.
+
+In cases where the action that causes the callback or constructor to be invoked
+originates with the browser, not being caused by any JS code, then the callback
+will have the same context as when the API that registered it was called. For
+worklets, only JS code inside the worklet counts as causing the action. In these
+cases, there is no other AsyncContext snapshot to consider using besides the one
+which was active when the API was called, so the generic WebIDL solution works.
+
+- [`navigator.mediaSession.setActionHandler()`](https://w3c.github.io/mediasession/#dom-mediasession-setactionhandler)
+  method [\[MEDIASESSION\]](https://w3c.github.io/mediasession/)
+- [`navigator.geolocation.watchPosition()`](https://w3c.github.io/geolocation/#dom-geolocation-watchposition)
+  method [\[GEOLOCATION\]](https://w3c.github.io/geolocation/)
+- [`RemotePlayback`](https://w3c.github.io/remote-playback/#dom-remoteplayback):
+  [`watchAvailability()`](https://w3c.github.io/remote-playback/#dom-remoteplayback-watchavailability)
+  method [\[REMOTE-PLAYBACK\]](https://w3c.github.io/remote-playback/)
+- [`AudioWorkletGlobalScope`](https://webaudio.github.io/web-audio-api/#AudioWorkletGlobalScope):
+  [`registerProcessor()`](https://webaudio.github.io/web-audio-api/#dom-audioworkletglobalscope-registerprocessor)
+  method [\[WEBAUDIO\]](https://webaudio.github.io/web-audio-api/)
+- [`PaintWorkletGlobalScope`](https://drafts.css-houdini.org/css-paint-api-1/#paintworkletglobalscope):
+  [`registerPaint()`](https://drafts.css-houdini.org/css-paint-api-1/#dom-paintworkletglobalscope-registerpaint)
+  method [\[CSS-PAINT-API\]](https://drafts.css-houdini.org/css-paint-api-1/)
+
+In the case of custom elements, the actions that invoke the custom element
+methods are usually triggered from JS code calling a web API, e.g. removing a
+custom element with `el.remove()` will cause its `disconnectedCallback` method
+to be run. But in some cases custom element reactions can also run as a response
+to a user action, such as removing a custom element inside a text editing
+context. To avoid having a difference between these cases, we propose using the
+context in which the custom element was defined with
+[`customElements.define()`](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-customelementregistry-define)
+[\[HTML\]](https://html.spec.whatwg.org/multipage/). In the future it might be
+possible to add a way to get the snapshot in which el.remove() was invoked, if
+there are use cases.
+
+In the cases where the registration web API takes a constructor, the methods and
+getters of the returned object (e.g. the `processor` method in web audio
+worklets, or the `attributeChangedCallback` method of custom elements) also
+count as callbacks that these APIs invoke, and which should preserve the
+relevant context.
```

Comment:
This is somewhat related to the old issue https://github.com/whatwg/webidl/issues/701: the way in which these are specified today, by using raw JS machinery instead of a standardized Web IDL pattern, is annoying to get right and error-prone.

Certainly not a blocker, but if someone wanted to shave that yak while they were doing the web integration here, there would be much rejoicing.